Adjust paginator to use api page_count

mihaidogaru2537 opened the pull request:
<!--
    Thanks for opening a pull request (PR). Please read through
    these instructions to help us review and merge your change quicker.

    Replace ISSUE_NUMBER with the number of the issue related to what
    you're fixing followed by a description of your change. For example:

    Fixes #3588

    This patch adds a margin on the logout button to correct the layout.
-->

Fixes #10689 

Changes:

- Added the "page_count" number received by the request triggered from SEARCH_STARTED action into the "search" redux state.
- Mapped the search.page_count state to the SearchBase component's state.
- Adjusted Paginate component to receive one more prop called `receivedPageCount`. If this prop is passed down, then the Paginate component no longer compute the page count manually and it will use this value instead.

**Before**:

The paginator does not stop at page index 1200 although this is the last page the backend will return.
The paginator computes the total number of pages manually using the total number of items (received from API) and the total number of items per page (received from API).

![image](https://user-images.githubusercontent.com/61188230/207085213-bfe1dddd-50eb-4f23-922d-4da4d5e02901.png)

Clicking paginator's page index 1201 leads to an error message (as it should)

![image](https://user-images.githubusercontent.com/61188230/207085432-461cf670-2433-4aff-85a4-4a2d95dc4ac7.png)



**After**:

The paginator uses the maximum page number received from the API and stops at 1200.
The next button is disabled at page 1200.
Page x out of y message is updated according to the page_count received from the API.

![image](https://user-images.githubusercontent.com/61188230/207084201-4fae0b14-ba1e-45d0-b1b6-dcb8735cf595.png)

Manually changing the query param to 1201 still displays an error message.

![image](https://user-images.githubusercontent.com/61188230/207084454-db1b1866-f6ea-43e1-926a-f143be94c122.png)


<!--
    Read through this checklist to make sure your patch is ready for review.

    - Add a PR title that summarizes your patch
    - Make sure this PR relates to an existing open issue and there are no existing PRs open for the same issue.
    - Add tests to cover the changes added in this PR.
    - Check that the change works locally.
    - Add before and after screenshots (Only for changes that impact the UI).

    Thanks for your contribution!
-->

mihaidogaru2537 commented:
**Questions / Notes**

1. Do we need to add automated tests for this feature?
2. Is it good practice to wait until someone confirms that my work is actually solving the problem, before adding any automated tests?

This is my first ever contribution, thanks for reading!

diox commented:
Thanks for your contribution, it's looking good! See my answers to your questions below:

> Do we need to add automated tests for this feature?

Yes, we do require tests. One nice thing here is existing tests passing should prove other uses of the pagination are not affected by this change.

> Is it good practice to wait until someone confirms that my work is actually solving the problem, before adding any automated tests?

I guess it depends on how confident you are in your solution. It's totally fine to request feedback before adding tests if you're unsure this is the right approach. In this case I think it's good, you can request a full review from us once you're done adding tests. (@bobsilverberg , you might want to take a look too ?)

bobsilverberg commented on src/amo/reducers/search.js:
```diff
@@ -125,6 +129,7 @@ export default function search(
         results: payload.results.map((addon) =>
           createInternalAddon(addon, state.lang),
         ),
+        pageCount: payload.pageCount,
```
In addition to this, we should probably set `pageCount` to `0` (or maybe to `null`, but `0` is probably fine) in both `SEARCH_STARTED` and `SEARCH_ABORTED` so that it gets reset each time a search is run, and doesn't carry over the value from the last search.

bobsilverberg reviewed (changes requested):
This is really nice work, thanks @mihaidogaru2537! I have made a couple of suggestions, and, as @diox indicated, tests should be added for any code you've added/changed. Feel free to request another review after the suggestions have been addressed and tests have been written/updated.

mihaidogaru2537 commented on src/amo/components/Paginate/index.js:
```diff
@@ -174,7 +178,7 @@ export class PaginateBase extends React.Component<InternalProps> {
           {i18n.sprintf(i18n.gettext('Page %(thisPage)s of %(totalPages)s'), {
             thisPage,
-            totalPages: getPageCount({ count, perPage }),
+            totalPages: receivedPageCount || getPageCount({ count, perPage }),
```
Fixed in my 2nd commit.

bobsilverberg commented on tests/unit/amo/components/TestPaginate.js:
```diff
@@ -186,6 +186,32 @@ describe(__filename, () => {
     );
   });
 
+  describe('when receivedPageCount is passed', () => {
```
This test looks good, but it doesn't really need to be wrapped in its own `describe`. You could just rename the test to `it('uses receivedPageCount for paginator links when passed', () => {`

mihaidogaru2537 replied:
Done, thanks for the feedback!

bobsilverberg commented:
> Thank you all for your feedback. I added the code changes needed for the refactor in my 2nd commit.

Thanks @mihaidogaru2537, these changes look good.

> Then, in my last commit I added the automated tests.
> 
> The tests are as follows:
> 
> 1. tested the Paginate component such that when it receives the receivedPageCount prop, it will use it.

This one looks good. 

> 2. tested the search saga such that START_SEARCH and ABORT_SEARCH will reset the receivedPageCount state to 0.

I think you mean here that you updated the tests for the search *reducer*, not the search *saga*. These tests look good as well, but you should also update the test for `SEARCH_LOADED` so that it receives a `pageCount` from the response and stores it in the store.

> 
> This is my first time writing front end automated tests and I am not 100% sure whether what I tested is enough. I was thinking maybe I should test these too:
> 
> 1. the test_search from /amo/api, but from what I have seen there, only the input is tested there and we do not alter the input.

No changes were made to the search api code, so no changes are needed for those tests.

> 
> 2. the test_search from /amo/sagas, but again, it seems like only the actions returned by the saga are tested, not their content (which we modified)

Because the saga is now taking data from the API and transforming it, i.e., `page_count` is being dispatched in the action as `pageCount`, a test should be added for that. It doesn't necessarily need a new test, it could be added to the test for `it('searches the API for add-ons'`, but we should verify that the action looks the same as what we expect. You can see an example for doing a test like that at https://github.com/mozilla/addons-frontend/blob/master/tests/unit/amo/sagas/test_abuse.js#L54-L74.

> 
> 3. the Search component, we altered its body but it seems like there is no already existing test targeting this component.

This component is tested in `TestSearchPage.js`, starting at https://github.com/mozilla/addons-frontend/blob/master/tests/unit/amo/pages/TestSearchPage.js#L362. You should be able to find a test in there to either update, or duplicate and make changes to to cover the changes you made.

bobsilverberg reviewed (changes requested):
Thanks for the updates @mihaidogaru2537, this is looking really good. I have added some more comments.

mihaidogaru2537 commented:
> I think you mean here that you updated the tests for the search reducer, not the search saga. These tests look good as well, but you should also update the test for SEARCH_LOADED so that it receives a pageCount from the response and stores it in the store.

I added the "[Update SEARCH_LOADED reducer test](https://github.com/mozilla/addons-frontend/pull/11943/commits/55ca021da2f825e73937d088bef06abc81ad1e3a)" commit to take care of this.


> Because the saga is now taking data from the API and transforming it, i.e., page_count is being dispatched in the action as pageCount, a test should be added for that. It doesn't necessarily need a new test, it could be added to the test for it('searches the API for add-ons', but we should verify that the action looks the same as what we expect. You can see an example for doing a test like that at https://github.com/mozilla/addons-frontend/blob/master/tests/unit/amo/sagas/test_abuse.js#L54-L74.

Added "[Test search saga dispatches searchLoad action](https://github.com/mozilla/addons-frontend/pull/11943/commits/1fd3775de887b78d5c5d233ee68bcad747c21d3d)" commit to solve this

> This component is tested in TestSearchPage.js, starting at https://github.com/mozilla/addons-frontend/blob/master/tests/unit/amo/pages/TestSearchPage.js#L362. You should be able to find a test in there to either update, or duplicate and make changes to to cover the changes you made.

Added "[Test Search component uses pageCount for its paginator](https://github.com/mozilla/addons-frontend/pull/11943/commits/c0e147b9752dd1ae4df01422b3736a2d75359bc0)" commit for this one.
I am having a little bit of doubts here, so let me know if I should change anything here. (I was thinking maybe we should remove the `count `variable from `dispatchSearchResults `entirely) (but the `count `is still passed down into the props so maybe we should let it there for now)


Sorry for the delay, I somehow missed your reply where you requested all these changes.
Thanks again for the feedback and let me know if we should edit anything else.

bobsilverberg reviewed (approved):
This looks good, thanks for all of your efforts in getting this done. My apologies for taking so long to do this final review. Well done!
